### PR TITLE
splits each tool to have its own gopath

### PR DESCRIPTION
**Pull request by crxpandion**
This implements https://github.com/twitchtv/retool/issues/17. 
This branch makes each tool have its own gopath namespaced by the last folder in the package name. 

Each tool will then have its own copy of all dependencies. A follow up to this will be to add a way to checkout specific versions of its dependencies and attach that the the tools' manifest. This way we should be able to be much smarter about when we should fetch from the tools or dependency's origin.

This branch also fixes a bug in the `SetEnvVar` function when multiple calls are made on the same `cmd`

### CODE REVIEW

File: do.go [hunk 1]
```diff
@@ -22,18 +22,21 @@ func setPath() error {
 // GOBIN is set so gometalinter will use it to decide where to put its vendored
 // linters with the gometalinter --install command, and so that it prefers the
 // binaries built in _tools/bin when executing linters.
-func setGoEnv() error {
+func setGoEnv(tools []*tool) error {
```
Comment (spenczar) [inline do.go]: please update the comment at L16 to describe what this does now

File: tool.go [hunk 2]
```diff
@@ -18,7 +20,12 @@ type tool struct {
 }
 
 func (t *tool) path() string {
-	return filepath.Join(toolDirPath, "src", t.Repository)
+	return filepath.Join(t.gopath(), "src", t.Repository)
+}
+
+func (t *tool) gopath() string {
+	s := strings.Split(t.Repository, "/")
```
Comment (spenczar) [inline tool.go]: Use `path.Base(t.Repository)` here

File: tool.go [hunk 3]
```diff
@@ -27,30 +34,35 @@ func (t *tool) executable() string {
 
 func setEnvVar(cmd *exec.Cmd, key, val string) {
```
Comment (spenczar) [inline tool.go]: We should maybe use this for the `do` command when it sets GOBIN/GOPATH/PATH

File: tool.go [hunk 4]
```diff
@@ -132,9 +145,16 @@ func download(t *tool) error {
 
 func install(t *tool) error {
 	log("installing " + t.Repository)
+
+	err := ioutil.WriteFile(path.Join(t.gopath(), ".gitignore"), gitignore, 0664)
```
Comment (spenczar) [inline tool.go]: We'll also need to write a `.gitignore` into `_tools/.gitignore` to ignore `_tools/bin`

**Comment by crxpandion** [comment]
I see two options:
1. Version retool and record the version in the manifest. Based on the manifest version either expect this new paradigm or the old one. Prompt the user to upgrade if the manifest is old, but continue to work.
2. Detect we are in the old paradigm implicitly and prompt the user to upgrade, but again continue to work.

**Comment by spenczar** [comment]
Marking the retool version in the manifest just seems like a good idea in general. Let's go with that direction.

**Comment by spenczar** [comment]
On further thought, we can't do this any time very soon. The issue will be for people who have an old version of retool who work on a project that was synced with this new version. The old version of retool will not understand that tools are already vendored, and will double-vendor them.

This is too traumatic of a change for retool. We can't break peoples' workflows that harshly.

This might be okay if we had clear, semantic versioning baked into the manifest and lock file. That way, retool could examine the version in the manifest/lock, compare it against its own compiled version, and emit a warning or error if it's a full major version out of date.

For now, this will have to be on hold. I'd want that versioned retool to disseminate broadly (say, for 6 or 9 months) before using it too heavily.

**Pull request closed without merging**